### DISCUSSION TITLE
Re-focusing on our open source community

### DISCUSSION THREAD

**Discussion by nbolton** [deskflow/deskflow#7476] [General]
Where do I start? OK, we messed up our open source developer community.

The last 10 years have been a challenging journey. We tried something new with our commercial venture, Synergy 2 and it flopped. We then bounced back with Synergy 3 which has been a hit with paying customers, but it's left our open source community asking: "What about Synergy 1?"

I'm glad to say that development of Synergy 1 is still important to us. Crucial, in fact. This is not just because Synergy 3 depends on the open-source code of Synergy 1 (it uses the Core component; server and client) but also because we are an open-source company at our foundation. Synergy 3 is just a friendly wrapper around Synergy 1, where the real magic still happens. The Synergy 3 GUI is proprietary now, but that may change.

We have made the mistake of neglecting our open-source community while we focused on the new GUI. As a result, many new Synergy-derived forks have popped up over the last decade, each with their re-imagining of not only the Synergy logo but an investment of significant effort into improving the Synergy-like source code.

Barrier, sadly, came and went. Though it is dead now, its message rings clear: we weren't managing the Synergy community properly. Then came Barrier's descendant, Input Leap (a Synergy-derivative). I would like to take a moment here to commend the developers who have invested significant effort into all Synergy-derived forks, efforts that have led to significant improvements in Synergy 1 Community Edition. Developers such as Peter Hutterer (@whot), the primary developer of Wayland support for Synergy-like apps, have made leaps that benefit not only Synergy but all projects that are now able to use [`libei`](https://gitlab.freedesktop.org/libinput/libei) and [`libportal`](https://github.com/flatpak/libportal) to enable remote control and input capture on non-X desktops (i.e. Wayland, etc). Many others deserve credit, such as @ofourdan and @p12tic for their hard work on improving the forks.

Anyway, I want you all to know that we're back and we care about rebuilding our open-source community. We spent too long focusing _only_ on our customers (who are still important because they allow us to pay our developers) and realized that we need to prioritize the Synergy community.

I hope those who have left our community will take a moment to think about rejoining us.

p.s. If you are a former community member, please leave a comment on this discussion along the lines of "I left because [reason] and [plan/do not plan] on rejoining because...", this would help us to make changes.

**Comment by ahesford** [comment]
You lost me when I paid $10 for Synergy 2 and you shortly thereafter walked it back to beta. Seeing a promise that my license would be carried forward when you re-released v2, I watched for years as it turned into vaporware.

Synergy 2 didn't "flop", Symless bungled it.

**Reply by nbolton** [reply]
Thanks for the feedback. I am sorry about the mistakes we made with Synergy 2. Did you know that you were given a free upgrade to Synergy 3? You can get it simply by logging into your account. Let me know if you need help with that.

While I appreciate your feedback as a customer, I was looking more for feedback around our open source community rather than the licensed offering. Does your feedback still apply to the open source Synergy development community or was it purely in relation to licensing?

**Reply by ahesford** [reply]
I don't have any particular experience with Synergy as an open-source project. In general, it's hard not to be cynical about open-source projects that serve as the core for paid software. There is always a tension between the nature of open source and the interests of the company that builds a product atop it. Open source undercuts many of the avenues by which a corporation can profit from its software. If anybody can just grab a copy of your software for free, what do you offer? Support? Proprietary additions? These can be hard to build a business around.

Often, it feels like companies that offer open-source software do so not in the spirit of collaboration and contribution, but in the hopes that their user community will improve the product for them. Signs of this include cumbersome build policies or a tendency to vendor dependencies in a way that makes the software difficult to package, but is convenient for the company. (This is not directed at Symless; I haven't looked closely at the Synergy source repository, but a cursory glance suggests that it is a standard, manageable meson project.) Many times, companies seem to conclude that this is a net negative and try to put the genie back in the bottle. Several high-profile software companies have changed liceneses on their popular products to decrease source availability. That usually works out well for nobody, although the tendency seems to be the emergence of a dominant community-driven fork and the eventual irrelevance of the original corporate project (*cough* ZFS *cough*).

**Reply by nbolton** [reply]
Thanks. Great insight! 

I'm still curious whether you used your free upgrade or got a refund. It's not too late for either option.

I have been pouring a lot of work into trying to make Synergy easy to build lately, though it was already pretty easy to build (except for on Windows where building anything can be horrible). Most of the effort seems to go into making Windows easier to build but I don't think I'm there yet. It's a CMake project by the way, we're just playing around with Meson (with the idea of potentially migrating to it from CMake in future).

I definitely see your skepticism about the motivations for a corporation having an open source project; for free work, not for collaboration. My motive for running a company is to have professional developers improve Synergy, but by nature of being a company, some of that money must be focused on customers because that's where the money comes from.  I'm sure Red Hat has the same dynamic. I do have plans to separate the upstream from Synergy and give control to the community (like Fedora), so that the company is adapting to the open source upstream and not the other way around, but this will be a long journey and must be done properly. Interop between Synergy-derived (eg Input Leap) and Synergy-like (Lan Mouse) is something that we discussing with each other.

If you have any questions, feel free to ask 🙂

**Comment by gparyani** [comment]
It took ten years and a major flop, but I'm glad you did end up realizing that the open source community is important.

I have to tell you something. Back in 2014, I was working on a university research project on parallel computing where we had a set 15 Raspberry Pi computers wired to a 5x3 wall of monitors. Synergy was extremely useful to us at the time as it allowed us to control the entire wall using only a single keyboard and mouse. It was very nice that someone had compiled the project for ARM32 and made it available on the APT repositories, and the university had planned to give away a donation to the project for its great help to us. However, when they were about to do it, that was *right* when you'd decided to start charging for access to compiled binaries. The administrators saw the [FAQ you'd put up at the time](https://web.archive.org/web/20141107195719/http://synergy-project.org/download/) which contained many insults (which I won't quote here) and decided they didn't want to support such a developer. You did [later retract the insults](https://web.archive.org/web/20141114011045/http://synergy-project.org/download/), but by that time, they'd already made their decision.

That decision was quite disheartening, as it forced us to compile the source on a single-core processor with only 512 MB RAM just to get the latest changes on the computers - another major factor in the university's subsequent decision to not give any money. The subsequent decision to go commercial proprietary with the software made it impossible for us to use it (we wanted everything to be open source). I since ceased working on the project, so I don't know what they did next, but it did stymie the team greatly.

I understand that a lot was going through your mind at the time because your main sponsor had pulled out, leaving you with no choice but to look for ways to monetize your software, and it's natural for any of us to rant under stress. It was also a case of unfortunate timing that the university admins had seen the prior version of the FAQ with the rants and based their decision on that - had they seen the updated version, it's likely they'd have come to a different decision.

I came back here because I saw a shared article on the renaming of the project to Deskflow and the refocusing on open source developers. I am very glad you're coming back to them. As I'm no longer involved with the project (and from what I've heard on the grapevine it's since been dismantled), I'm probably not going to be involved in the community, but I'm glad the refocusing happened at all, even if it took ten years. It'll definitely help if they or another entity decide to build something similar.

**Reply by nbolton** [reply]
Thank you for sharing your story, and I'm sorry to hear that things turned out that way.

> The administrators saw the [FAQ you'd put up at the time](https://web.archive.org/web/20141107195719/http://synergy-project.org/download/) which contained many insults (which I won't quote here) and decided they didn't want to support such a developer. You did [later retract the insults](https://web.archive.org/web/20141114011045/http://synergy-project.org/download/), but by that time, they'd already made their decision.

Excellent research. I had a lot of growing up to do 10 years ago. It's regrettable that we missed out on this opportunity and I'm sorry to hear that your project was affected.

> decision to go commercial proprietary with the software

FWIW, we continued to develop the open-source GPL2 version which has always been publicly available (it was under [symless/synergy-core](https://github.com/symless/synergy-core) on GitHub) before that repo was moved to Deskflow.

> which I won't quote here

I'm keen to learn from my mistakes. If you prefer real-time chat, I'm also available on [Discord](https://discord.com/invite/xBFv6j7) and IRC (`#deskflow` on [Libera.Chat](https://libera.chat/)).

**Comment by zhurai** [comment]
Apologies for putting this here if it's not the best area, but if both input-leap and deskflow (and lan mouse) are planning to be interop with each other.

What specific differences will there be between the open source projects? I believe I read somewhere else that InputLeap is on Qt5 for compatibility reasons but Deskflow is on Qt6? Are there other differences? 
I believe I have seen some push requests copied and adjusted from one to another to help with certain bugs that happened on both (for good reason but potentially worth mentioning)

This is mostly of a semi personal reason but I tend to just install whatever works/has the features I like, but if they're both the same then it becomes somewhat random what I'd download I suppose. 

--- 

(random rambling context) 

As for me in general, I'm an old(?) user that also supported by buying a license back in 2017, but I tend to go for whatever has the features that I need as the first priority, and I wasn't really involved with the open source/development side of things back then.  I had to leave Synergy and it's related projects for awhile, as I switched to Wayland Linux (Sway) and was unable to control my Macbook from my main machine.    There was enough disruptions that I ended up using a secondary computer for other reasons, but also to run X11 to control the Macbook controlled using rkvm... 

Eventually I did find and test lan mouse too but ran into some weird issues with controlling my mac + some weird instabilities in the netcode for both rkvm/lanmouse that I don't seem to get with the Synergy branch ones (Synergy/deskflow/inputleap).   Eventually, I found that KDE also supported the related protocols... so I have been testing/finetuning KDE switching over from sway just to use deskflow/inputleap, but I've been in a weird tossup between the two projects (no differentiator between which to install)

[At this moment I'm on inputleap because the aur package wasn't building as the tests were failing, though the maintainer has seem to have figured out the issue causing the test failures... which means I'm not sure of what I would be missing out between using one or the other... setting either of these up on my mac w/ the commands to prevent the OS from reporting the .dmg is "damaged" among other annoyances makes it a drag to "just use/try both"]

**Reply by sithlord48** [reply]
The issues with [wayland](https://github.com/deskflow/deskflow/discussions/7499) are documented.

**Reply by zhurai** [reply]
> The issues with [wayland](https://github.com/deskflow/deskflow/discussions/7499) are documented.

Hello, That wasn't what I was asking about/currently having issues with! :slightly_smiling_face: 

I'm aware of that discussion - I've also been aware of the status that it didn't work on wlr based compositors (e.g. sway/hyprland | as it has not been implemented completely on their end) since around when input-leap initially implemented it, but prior to it being communicated in the issues that it was for gnome only.  It's why I was trying rkvm/lan-mouse for the sway portion, but was running into some weird netcode related issues that I wasn't getting on synergy based.  

I didn't end up going with Gnome because I dislike that as a choice, but decided to switch over to KDE once I found it was confirmed that it worked with 6.1+ KDE  (still some rough parts in comparison to my sway setup though).

But yes I'm very well aware, otherwise I would not have tested rkvm/lan-mouse as well as switched back to X11 for i3 for several months prior to trying+using KDE, thanks!  :slightly_smiling_face: 

(I'm just trying to find out the differences between the two projects in this case so I can make an informed decision between the two... otherwise I get into a sort of analysis paralysis on choosing between two choices when I don't know the differences between the choices!)

**Reply by nbolton** [reply]
> I read somewhere else that InputLeap is on Qt5 for compatibility reasons but Deskflow is on Qt6? Are there other differences?

Things move quickly! This is actually no longer the case as IL dropped Qt5 a few days ago. 

The major differences between our projects now seem to be in project values (which we display on our README) and GUI style. For example, we require all PRs to go through peer review, whereas IL does not. There are pros and cons of both approaches.

> supported by buying a license back in 2017

Thank you! :heart: 

> lan mouse too but ran into some weird issues with controlling my mac

Cool project. I think it's a great idea; rewriting Synergy in Rust is an extremely ambitious project and @feschber has made unbelievable progress recently. There are pros and cons of starting again, primarily, you don't have the 25-ish years of tech debt that Synergy/Deskflow/Barrier/Input Leap has (the first version was in 2001) but you also don't have the 25-ish years of bug fixes and experience... so which one will produce the best user experience, IDK! What I do know is that we will learn things from each other and Lan Mouse will be _blazingly fast_ :wink: 

> reporting the .dmg is "damaged"

Deskflow is new so packaging issues will be around for a few months until we get our stuff together.

> (I'm just trying to find out the differences between the two projects in this case so I can make an informed decision between the two... otherwise I get into a sort of analysis paralysis on choosing between two choices when I don't know the differences between the choices!)

TL;DR: We have different bug fixes and different ideological positions... and a different looking GUI. Other than that the two projects will be very similar. Some people hope we will merge the projects since they're so similar (I hope this too!)

**Comment by nbolton** [comment]
Thanks for all of your responses, I'll close the discussion but if anyone has any further thoughts I'd love to hear them.

Please feel welcome to join us on Matrix: https://matrix.to/#/#deskflow:matrix.org
